lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,923 @@
+                                     validate_certs=validate_certs, metadata=meta, files=files)
+
+    @staticmethod
+    def from_path(path, validate_certs, force, parent=None):
+        info = {}
+        for file_name, property_name in CollectionRequirement._FILE_MAPPING:
+            file_path = os.path.join(path, file_name)
+            if os.path.exists(file_path):
+                with open(file_path, 'r') as file_obj:
+                    try:
+                        info[property_name] = json.load(file_obj)
+                    except ValueError:
+                        raise AnsibleError("Collection file at '%s' is not a valid json string." % file_path)
+
+        if 'manifest' in info:
+            meta = info['manifest']['collection_info']
+        else:
+            display.warning("Collection at '%s' does not have a MANIFEST.json file, cannot detect version."
+                            % to_text(path))
+            parent_dir, name = os.path.split(path)
+            namespace = os.path.split(parent_dir)[1]
+            meta = {
+                'namespace': namespace,
+                'name': name,
+                'version': '*',
+                'dependencies': {},
+            }
+
+        namespace = meta['namespace']
+        name = meta['name']
+        version = meta['version']
+
+        files = info.get('files', {}).get('files', {})
+
+        return CollectionRequirement(namespace, name, path, None, [version], version, force, parent=parent,
+                                     validate_certs=validate_certs, metadata=meta, files=files, skip=True)
+
+    @staticmethod
+    def from_name(collection, servers, requirement, validate_certs, force, parent=None):
+        namespace, name = collection.split('.', 1)
+        galaxy_info = None
+        galaxy_meta = None
+
+        for server in servers:
+            collection_url_paths = [server, 'api', 'v2', 'collections', namespace, name, 'versions']
+
+            is_single = False
+            if not (requirement == '*' or requirement.startswith('<') or requirement.startswith('>') or
+                    requirement.startswith('!=')):
+                if requirement.startswith('='):
+                    requirement = requirement.lstrip('=')
+
+                collection_url_paths.append(requirement)
+                is_single = True
+
+            collection_uri = _urljoin(*collection_url_paths)
+            try:
+                resp = json.load(open_url(collection_uri, validate_certs=validate_certs))
+            except urllib_error.HTTPError as err:
+                if err.code == 404:
+                    continue
+                raise
+
+            if is_single:
+                galaxy_info = resp
+                galaxy_meta = resp['metadata']
+                versions = [resp['version']]
+            else:
+                versions = []
+                while True:
```

Comment:
```suggestion
                while resp['next'] is not None:
```
?

Reply:
Won't work because we need to run the loop at least once. If the versions doesn't contain a 2nd page then `resp['next']` is None and we would never pick up any versions.